### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by heitorcolangelo**

### CODE REVIEW

**Comment by heitorcolangelo** [comment]
Hi Bill! I'm fine, thanks. How are you and Amanda?
Thank you for reviewing my code challenge.
It has been really rush this month here and I didn't have time to pay enough attention and delivery the challenge that I wanted to.
I saw some of the questions and I'll be answering all of them as soon as possible. I will also include the missing screenshots and libs explanation.
Thanks again!

File: app/build.gradle [hunk 1]
```diff
@@ -1,33 +1,97 @@
...
-
 apply plugin: 'kotlin-android-extensions'
+apply from: "$rootDir/tools/ktlint.gradle"
+
+final versions = rootProject.ext
 
 android {
-    compileSdkVersion 26
+
+    dataBinding {
```
[the remark this answers is not in the document]
Comment (heitorcolangelo) [reply app/build.gradle]: No special reason. I'm only more confident in using databinding over kotlin extensions.

File: app/build.gradle [hunk 2]
```diff
@@ -1,33 +1,97 @@
...
+        enabled true
+    }
+
+    compileSdkVersion versions.compileSdk
```
[the remark this answers is not in the document]
Comment (heitorcolangelo) [reply app/build.gradle]: I don't think so, if you think that the architecture should support project growth, this avoid declaring multiple versions of the same library between modules, you centralize the information about versions in one place. Also, you commonly check the dependecies at the beginning of the project, but the tendency is to do this less and less

File: app/build.gradle [hunk 3]
```diff
@@ -1,33 +1,97 @@
...
+        }
+
         release {
             minifyEnabled false
             proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
         }
     }
+
+    lintOptions {
+        abortOnError false
```
[the remark this answers is not in the document]
Comment (heitorcolangelo) [reply app/build.gradle]: Oh yeah, I'm sorry for that. I put it in the beginning of the project and forgot to remove it later. It compiles with no errors if you remove.

File: app/build.gradle [hunk 4]
```diff
@@ -1,33 +1,97 @@
...
+    compileOptions {
+        targetCompatibility 1.8
+        sourceCompatibility 1.8
+    }
 }
 
 dependencies {
+
+    implementation project(':repository')
     implementation fileTree(dir: 'libs', include: ['*.jar'])
```
[the remark this answers is not in the document]
Comment (heitorcolangelo) [reply app/build.gradle]: Yes, this should have been removed.

File: app/build.gradle [hunk 5]
```diff
@@ -1,33 +1,97 @@
...
-    implementation"org.jetbrains.kotlin:kotlin-stdlib-jre7:$kotlin_version"
-    implementation 'com.android.support:appcompat-v7:26.1.0'
-    implementation 'com.android.support.constraint:constraint-layout:1.0.2'
-    testImplementation 'junit:junit:4.12'
-    androidTestImplementation 'com.android.support.test:runner:1.0.1'
-    androidTestImplementation 'com.android.support.test.espresso:espresso-core:3.0.1'
+    kapt "com.android.databinding:compiler:3.0.1"
```
[the remark this answers is not in the document]
Comment (heitorcolangelo) [reply app/build.gradle]: This should be the same as gradle version, I've forgot to change.

File: app/src/main/kotlin/com/heitorcolangelo/movies/CustomApplication.kt [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+import com.heitorcolangelo.movies.repository.RepositoryProvider
+import timber.log.Timber
+
+class CustomApplication : Application() {
+
+    override fun onCreate() {
+        super.onCreate()
+        if (BuildConfig.DEBUG) Timber.plant(Timber.DebugTree())
+        RepositoryProvider.init(this)
+        GenreManager.genreList()
```
[the remark this answers is not in the document]
Comment (heitorcolangelo) [reply app/src/main/kotlin/com/heitorcolangelo/movies/CustomApplication.kt]: The idea is to get the genres as soon as the application is created. As the remote data is requested through the managers classes, I tried to keep the same pattern. The idea here is that we had the genres before the movies list.

File: app/src/main/kotlin/com/heitorcolangelo/movies/ui/adapter/PaginationAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
+        super.onAttachedToRecyclerView(recyclerView)
+        recyclerView.clearOnChildAttachStateChangeListeners()
+        recyclerView.addOnChildAttachStateChangeListener(childAttachListener())
+    }
+
+    override fun saveInstanceState(): Bundle {
+        val bundle: Bundle = super.saveInstanceState()
+        bundle.putBoolean(HAS_NEXT_KEY, hasLoadingItem)
+        bundle.putBoolean(HAS_ERROR_KEY, hasError)
```
[the remark this answers is not in the document]
Comment (heitorcolangelo) [reply app/src/main/kotlin/com/heitorcolangelo/movies/ui/adapter/PaginationAdapter.kt]: That's a mistake, the progressBar visibility should be handled on the activity only on the first call. Sorry for that =/

File: app/src/main/kotlin/com/heitorcolangelo/movies/ui/adapter/ViewHolder.kt [hunk 8]
```diff
@@ -0,0 +1,13 @@
+package com.heitorcolangelo.movies.ui.adapter
+
+import android.databinding.ViewDataBinding
+import android.support.v7.widget.RecyclerView
+import com.heitorcolangelo.movies.BR
+
+class ViewHolder(val binding: ViewDataBinding) : RecyclerView.ViewHolder(binding.root) {
+
+    fun bind(`object`: Any, index: Int) {
```
[the remark this answers is not in the document]
Comment (heitorcolangelo) [reply app/src/main/kotlin/com/heitorcolangelo/movies/ui/adapter/ViewHolder.kt]: I commonly use this in my projects, The idea was to create transitions with shared elements, turns out I didn't have enough time.

File: app/src/main/kotlin/com/heitorcolangelo/movies/ui/dialog/LoadingFragment.kt [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+            if (!fInstance.isVisible) {
+                fInstance.show(activity.supportFragmentManager, LOADING_FRAGMENT)
+            }
+
+            return fInstance
+        }
+
+        fun hide(fragmentInstance: LoadingFragment?): LoadingFragment? {
+            fragmentInstance?.dismiss()
+            return null
```
[the remark this answers is not in the document]
Comment (heitorcolangelo) [reply app/src/main/kotlin/com/heitorcolangelo/movies/ui/dialog/LoadingFragment.kt]: It's just to keep the pattern in the BaseActivity
```kotlin
fun showLoading() {
    loadingFragment = LoadingFragment.show(this, loadingFragment)
}

fun hideLoading() {
    loadingFragment = LoadingFragment.hide(loadingFragment)
}
```

File: app/src/main/kotlin/com/heitorcolangelo/movies/ui/feature/medialist/MediaListActivity.kt [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    private fun showMediaDetails(media: Media) {
+        startActivity(MediaDetailActivity.intent(this, media))
+    }
+
+    private fun loadMediaPage(page: Int) {
+        showLoading()
+        subscriptions.add(MediaListManager.upComing(page)
+                .subscribe(this@MediaListActivity::onMediaSuccess, this@MediaListActivity::onMediaError))
```
[the remark this answers is not in the document]
Comment (heitorcolangelo) [reply app/src/main/kotlin/com/heitorcolangelo/movies/ui/feature/medialist/MediaListActivity.kt]: Yes, sorry for that =/

File: app/src/main/res/layout/activity_media_details.xml [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+    <android.support.design.widget.CoordinatorLayout
+        android:layout_width="match_parent"
+        android:layout_height="match_parent">
+
+        <android.support.design.widget.AppBarLayout
+            android:id="@+id/app_bar_layout"
+            android:layout_width="match_parent"
+            android:layout_height="224dp"
+            android:layout_marginTop="-24dp"
```
[the remark this answers is not in the document]
Comment (heitorcolangelo) [reply app/src/main/res/layout/activity_media_details.xml]: In the '-24dp' case, was to properly adjust the behavior of the collapsing toolbar. The other ones I forgot to extract to a dimens property

File: repository/build.gradle [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+
+    implementation "com.orhanobut:hawk:$versions.hawk"
+
+    testImplementation "junit:junit:$versions.jUnit"
+    testImplementation "br.com.concretesolutions:requestmatcher:$versions.requestMatcher"
+    testImplementation "com.squareup.okhttp3:mockwebserver:$versions.okHttp"
+    testImplementation "org.mockito:mockito-core:$versions.mockito"
+    testImplementation "org.hamcrest:hamcrest-all:$versions.hamcrest"
+    testImplementation "org.jetbrains.kotlin:kotlin-stdlib-jre7:$versions.kotlin"
+    testImplementation "net.vidageek:mirror:$versions.mirror"
```
[the remark this answers is not in the document]
Comment (heitorcolangelo) [reply repository/build.gradle]: The idea was to have test. Didn't have time.

File: repository/src/main/kotlin/com/heitorcolangelo/movies/repository/GenreRepository.kt [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+import com.heitorcolangelo.movies.repository.data.remote.MediaService
+import com.heitorcolangelo.movies.repository.model.Genre
+import io.reactivex.Observable
+
+class GenreRepository internal constructor(
+        private val localData: LocalDataProvider.GenreData,
+        private val remoteData: MediaService) {
+
+    fun getMoviesGenres(language: String): Observable<List<Genre>> {
+        return if (localData.hasGenreList())
```
[the remark this answers is not in the document]
Comment (heitorcolangelo) [reply repository/src/main/kotlin/com/heitorcolangelo/movies/repository/GenreRepository.kt]: Yes, I made this knowing that I'm not updating them. The ideal was to implement some invalidation routine that retrieve again the genres list. But unfortunately, there was no time available to do so.

**Comment by arctouch-emanuelmachado** [comment]
Hi Heitor, thank you for answering our questions! We've completed your evaluation for this step of our hiring process, our HR will soon get back in touch with you. :)
Best regards!

**Pull request closed without merging**